File: golang/simulation/combat/army.go
package combat

import "math/rand"

const (
	MAX_STRENGTH   = 5
	SEIGE_STRENGTH = 3
)

type Army struct {
	Regulars int
	Elites   int
	Leaders  int
}

func (a *Army) Init(r, e, l int) {
	a.Regulars = r
	a.Elites = e
	a.Leaders = l
}

func (a *Army) GetStrength() int {
	units := a.GetUnits()
	if units > 5 {
		return 5
	} else {
		return units
	}
}

func (a *Army) Copy() *Army {
	return &Army{
		Regulars: a.Regulars,
		Elites:   a.Elites,
		Leaders:  a.Leaders,
	}
}

func (a *Army) GetUnits() int {
	return a.Regulars + a.Elites
}

func (a *Army) TakeHit(beseiging bool) {
	units := a.GetUnits()
	if units == 0 {
		return
	} else if units > MAX_STRENGTH {
		a.Regulars -= 1
	} else if beseiging {
		if units > SEIGE_STRENGTH {
			a.Regulars -= 1
		} else {
			a.downgradeOrHit()
		}
	} else {
		a.downgradeOrHit()
	}
}

func (a *Army) TakeHits(hits int, beseiging bool) {
	for i := 0; i < hits; i++ {
		a.TakeHit(beseiging)
	}
}

func (a *Army) CombatRoll() (combatResult []int) {
	combatResult = make([]int, 0, 5)
	for i := 0; i < a.GetStrength(); i++ {
		res := rand.Intn(6) + 1
		combatResult = append(combatResult, res)
	}
	return
}

func (a *Army) LeaderRoll(missedCombat []int, effect CombatCard) (finalResult []int) {
	finalResult = make([]int, len(missedCombat))
	for i := 0; i < len(missedCombat); i++ {
		res := rand.Intn(6) + 1
		finalResult[i] = res
	}
	return
}

func (a *Army) downgradeOrHit() {
	if a.Elites > 0 {
		a.Elites -= 1
		a.Regulars += 1
	} else {
		a.Regulars -= 1
	}
}

File: pages/api/models/army.ts
export interface Army {
  regulars: number;
  elites: number;
  leaders: number;
}

export class BaseArmy {
  regulars: number = 0;
  elites: number = 0;
  leaders: number = 0;

  constructor(regs: number, elts: number, ldrs: number) {
    this.regulars = regs;
    this.elites = elts;
    this.leaders = ldrs;
  }

  isArmyAlive(): boolean {
    return this.getCombatStrength() > 0;
  }

  getCombatStrength(): number {
    return this.elites + this.regulars;
  }

  takeHit(): boolean {
    if (!this.isArmyAlive()) {
      return false;
    }
    if (this.getCombatStrength() > 5) {
      this.regulars--;
    } else if (this.elites > 0) {
      this.elites--;
      this.regulars++;
    } else {
      this.regulars--;
    }
    return true;
  }

  takeHits(hits: number) {
    for (let i = 0; i < hits; i += 1) {
      if (!this.takeHit()) {
        // army is dead
        break;
      }
    }
    return;
  }
}